src/object/srv_obj_migrate.c

```diff
@@ -2459,13 +2484,22 @@ migrate_one_epoch_object(daos_epoch_range_t *epr, struct migrate_pool_tls *tls,
 
 	memset(&anchor, 0, sizeof(anchor));
 	memset(&dkey_anchor, 0, sizeof(dkey_anchor));
-	dc_obj_shard2anchor(&dkey_anchor, arg->shard);
+	if (tls->mpt_opc != RB_OP_UPGRADE) {
+		dc_obj_shard2anchor(&dkey_anchor, arg->shard);
+		orig_flags = DIOF_TO_LEADER | DIOF_WITH_SPEC_EPOCH |
+			     DIOF_TO_SPEC_GROUP | DIOF_FOR_MIGRATION;
+		unpack_arg.new_layout_ver = tls->mpt_layout_version;
+	} else {
+		orig_flags = DIOF_TO_LEADER | DIOF_WITH_SPEC_EPOCH |
+			     DIOF_FOR_MIGRATION;
```

Comment:
not quite follow, why without DIOF_TO_SPEC_GROUP flag for UPGRADED, and with DIOF_TO_SPEC_GROUP for non-UPGRADE?

Reply:
The spec flag is for rebuild enumeration. So normal rebuild only need reconstruct the data for the specific group, for example for EC_2P2G2, if 1th shard fails, it only need to enumerate and construct group 1, (with this spec flag).

But for upgrade (or layout fixing), the new shard reconstruct needs to fetch all data of the object, instead of only group1, for example for ec parity rotation upgrade, it will need to enumerate all dkeys from all group to know if this dkey belong to the shard being reconstructed.